File: Haskell/src/Modelos/Coordenada.hs
{-# LANGUAGE DeriveGeneric #-}

module Modelos.Coordenada (Coordenada (..), CoordenadaClass (..)) where

import GHC.Generics (Generic)
import Data.Aeson (ToJSON, FromJSON, encode, decode, eitherDecode)

class CoordenadaClass c where
  getMascara :: c -> Char
  getElemEspecial :: c -> Char
  getAcertou :: c -> Bool

  setElem :: c -> Char -> c
  setAcertou :: c -> c


data Coordenada = Coordenada
  { mascara :: Char
  , elemEspecial :: Char
  , acertou :: Bool
  } deriving (Show, Generic)


instance CoordenadaClass Coordenada where
    getMascara = mascara
    getElemEspecial = elemEspecial
    getAcertou = acertou

    setElem coordenada elemento = coordenada { elemEspecial = elemento }
    setAcertou coordenada = coordenada { acertou = True }

instance ToJSON Coordenada
instance FromJSON Coordenada

File: Prolog/src/Modelos/Coordenada.pl
:- module(coordenada, [
        getMascara/2, getElemEspecial/2, getAcertou/2, 
        setElem/3, setAcertou/2, setMascara/3
]).

:- dynamic coordenada/3.

% coordenada(Mascara, ElemEspecial, Acertou)

% Getters

getMascara(coordenada(Mascara, _, _), Mascara).
getElemEspecial(coordenada(_, ElemEspecial, _), ElemEspecial).
getAcertou(coordenada(_, _, Acertou), Acertou).

% Setters

setElem(coordenada(Mascara, _, Acertou), NovoElem, coordenada(Mascara, NovoElem, Acertou)).
setAcertou(coordenada(Mascara, ElemEspecial, _), coordenada(Mascara, ElemEspecial, true)).
setMascara(coordenada(_, ElemEspecial, Acertou), NovaMascara, coordenada(NovaMascara, ElemEspecial, Acertou)).
